File: 프로그래머스/JavaScript/Lv. 2/42586. 기능개발/기능개발.js
function solution(progresses, speeds) {
  const day = [];

  for (let i = 0; i < progresses.length; i++) {
    day.push(Math.ceil((100 - progresses[i]) / speeds[i]));
  }

  let cnt = 0;
  let maxDay = day[0];
  let answer = [];

  for (let i = 0; i < day.length; i++) {
    if (day[i] <= maxDay) {
      cnt++;
    } else {
      answer.push(cnt);
      cnt = 1;
      maxDay = day[i];
    }
  }

  answer.push(cnt);
  return answer;
}

File: 프로그래머스/Java/Lv. 2/42586. 기능개발/기능개발.java
import java.util.ArrayDeque;
import java.util.ArrayList;

class Solution {
	public int[] solution(int[] progresses, int[] speeds) {
		ArrayList<Integer> result = new ArrayList<Integer>();

		ArrayDeque<Integer> day = new ArrayDeque<>();

		for (int i = 0; i < progresses.length; i++) {
			int d = (int) Math.ceil((100 - progresses[i]) / (double) speeds[i]);
			day.addLast(d);
		}

		while (!day.isEmpty()) {
			int now = day.pollFirst();
			int count = 1;

			while (!day.isEmpty() && day.peekFirst() <= now) {
				day.pollFirst();
				count++;
			}

			result.add(count);
		}

		int[] answer = new int[result.size()];

		for (int i = 0; i < answer.length; i++) {
			answer[i] = result.get(i);
		}

		return answer;
	}
}